rfcs/v3/supergraph-client.md

```diff
@@ -0,0 +1,202 @@
+* I want to apply validation logic before an existing mutation in the supergraph
+* I want to run complex authorization logic before applying a mutation
+* I want to run a saga/workflow of business logic interleaved with queries/mutations
+
+By allowing code-based connectors to call back into the Supergraph, customers could address these use cases by simply exposing new Open DD commands that are backed by code that composes together queries to the Supergraph as necessary.
+
+Currently connectors sit downstream from the v3-engine, which exposes the entire Supergraph as a GraphQL API. Connectors yield data upstream to the engine, but have no facility to call back to the engine.
+
+Theoretically any code connector could call back into the Supergraph right now just by importing a GraphQL client library and talking to the Supergraph like any other API client. However, there are three main issues that make this difficult:
+
+* Addressing: how does the connector get the URL to the Supergraph?
+* Authentication/Authorization: How does the connector authenticate to the Supergraph, and what controls what parts of the Supergraph it has access to?
+* Supergraph Schema Building: How does the connector acquire the GraphQL schema of the Supergraph?
+
+## Addressing
+Connectors are deployed seperately to the Supergraph, which means that the URL to the Supergraph is not available at deployment time. This is particularly true when you consider that many Supergraph builds can reuse the same deployed connector, yet one would wish for the connector to call back into the same Supergraph build that caused the connector to invoked in the first place.
+
+In practice this means that the v3-engine call to the connector would need to provide the connector with the address to the Supergraph instance that is invoking it. This could be achieved via argument binding in the `DataConnectorLink` in Open DD, in a similar fashion to how header passthrough is performed.
+
+```yaml
+kind: DataConnectorLink
+version: v1
+definition:
+  url: ...
+  schema: ...
+  argumentPresets:
+    - argumentName: supergraphClient
+      value:
+        supergraphClient:
+          type: GraphQL # Only option for now, maybe in the future we could offer other types, eg REST
+```
+
+The connector itself would expose NDC functions/procedures that would accept a SupergraphClient scalar type of JSON representation:
+
+```yaml
+scalar_types:
+  SupergraphClient:
+    representation: json
+  Bool:
+    representation: boolean
+procedures:
+  name: myFunction
+  arguments:
+    supergraphClient:
+      type:
+        named: SupergraphClient
+  resultType:
+    named: Bool
+```
+
+The engine would send a JSON object with the URL to the current Supergraph build's GraphQL API (when running locally, this would just be a localhost engine URL):
+
+```json
+{ "supergraphUrl": "https://settling-impala-7722-50bb46e1d5.ddn.hasura.app/graphql" }
```

Comment:
We need to work out how this would work in the case where there is a custom domain. Engine is only aware of the canonical build urls as custom domains are handled by the gateway. Also, we plan to introduce path based routing for build urls instead of hostname based, and gateway is going to be handling that as well.

A solution could be for the gateway to pass the actual hostname to engine as a separate header and then use that header as SupergraphURL.

Reply:
That could work. I think we may need an environment variable setting for it on v3-engine as well, for people self-hosting v3-engine without a gateway injecting headers, or for use when running locally. 

Perhaps the env var setting would offer three options:
* Get it from a header from the gateway
* Get it as a value from the env var itself (for self-hosting)
* No env var set (default to localhost)